src/main/java/net/snowflake/ingest/streaming/internal/ParquetRowBuffer.java

```diff
@@ -0,0 +1,595 @@
+    switch (logicalType) {
+      case DATE:
+        intVal = DataValidationUtil.validateAndParseDate(value);
+        break;
+      case TIME:
+        intVal =
+            DataValidationUtil.getTimeInScale(
+                    DataValidationUtil.getStringValue(value), column.getScale())
+                .intValue();
+        break;
+      default:
+        intVal = DataValidationUtil.validateAndParseInteger(value);
+        break;
+    }
+    return intVal;
+  }
+
+  private static BigInteger parseSb16(Object value, RowBufferStats stats, Integer columnScale) {
+    int scale = DataValidationUtil.validateAndParseInteger(columnScale);
+    BigDecimal bigDecimalValue = DataValidationUtil.validateAndParseBigDecimal(value);
+
+    // explicitly match the BigDecimal input scale with the Snowflake data type scale
+    bigDecimalValue = bigDecimalValue.setScale(scale, UNNECESSARY);
+
+    BigInteger intRep =
+        bigDecimalValue.multiply(BigDecimal.valueOf(Power10.intTable[scale])).toBigInteger();
+    stats.addIntValue(intRep);
+    return intRep;
+  }
+
+  private static byte[] getSb16Bytes(BigInteger intRep) {
+    byte[] bytes = intRep.toByteArray();
+    byte padByte = (byte) (bytes[0] < 0 ? -1 : 0);
+    byte[] bytesBE = new byte[16];
+    for (int i = 0; i < 16 - bytes.length; i++) {
+      bytesBE[i] = padByte;
+    }
+    System.arraycopy(bytes, 0, bytesBE, 16 - bytes.length, bytes.length);
+    return bytesBE;
+  }
+
+  @Override
+  void moveTempRowsToActualBuffer(int tempRowCount) {
+    data.addAll(tempData);
+  }
+
+  @Override
+  void clearTempRows() {
+    tempData.clear();
+  }
+
+  @Override
+  boolean hasColumns() {
+    return !fieldIndex.isEmpty();
+  }
+
+  @Override
+  Optional<ParquetChunkData> getSnapshot() {
+    List<List<Object>> oldData = new ArrayList<>();
+    data.forEach(r -> oldData.add(new ArrayList<>(r)));
+    return oldData.isEmpty()
+        ? Optional.empty()
+        : Optional.of(new ParquetChunkData(oldData, metadata));
+  }
+
+  @Override
+  Object getVectorValueAt(String column, int index) {
+    int colIndex = fieldIndex.get(column).getSecond();
+    return tempData.get(index).get(colIndex);
+  }
+
+  @Override
+  int getTempRowCount() {
+    return tempData.size();
+  }
+
+  @Override
+  void reset() {
+    data.clear();
+    this.rowCount = 0;
+    this.bufferSize = 0F;
+    this.statsMap.replaceAll(
+        (key, value) -> new RowBufferStats(value.getCollationDefinitionString()));
+  }
+
+  @Override
+  public void close(String name) {
+    this.fieldIndex.clear();
+    logger.logInfo(
+        "Trying to close parquet buffer for channel={} from function={}",
+        this.owningChannel.getName(),
+        name);
+  }
+
+  public static class ParquetChunkData {
+    final List<List<Object>> rows;
+    final Map<String, String> metadata;
+
+    public ParquetChunkData(List<List<Object>> rows, Map<String, String> metadata) {
+      this.rows = rows;
+      this.metadata = metadata;
+    }
```

Comment:
Should this be its own class with a corresponding docstring given how frequently it gets referenced elsewhere?